cmd/oras/root/manifest/index/update.go

```diff
@@ -0,0 +1,243 @@
+		}
+		if !descriptor.IsManifest(desc) {
+			return nil, fmt.Errorf("%s is not a manifest", manifest)
+		}
+		printUpdateStatus(status.IndexPromptFetched, manifest, string(desc.Digest), opts.Printer)
+		if descriptor.IsImageManifest(desc) {
+			desc.Platform, err = getPlatform(ctx, target, content)
+			if err != nil {
+				return nil, err
+			}
+		}
+		manifests = append(manifests, desc)
+		printUpdateStatus(status.IndexPromptAdded, manifest, string(desc.Digest), opts.Printer)
+	}
+	return manifests, nil
+}
+
+func mergeIndexes(ctx context.Context, manifests []ocispec.Descriptor, target oras.ReadOnlyTarget, opts updateOptions) ([]ocispec.Descriptor, error) {
+	for _, indexRef := range opts.mergeArguments {
+		printUpdateStatus(status.IndexPromptFetching, indexRef, "", opts.Printer)
+		desc, content, err := oras.FetchBytes(ctx, target, indexRef, oras.DefaultFetchBytesOptions)
+		if err != nil {
+			return nil, fmt.Errorf("could not find the index %s: %w", indexRef, err)
+		}
+		if !descriptor.IsIndex(desc) {
+			return nil, fmt.Errorf("%s is not an index", indexRef)
+		}
+		printUpdateStatus(status.IndexPromptFetched, indexRef, string(desc.Digest), opts.Printer)
+		var index ocispec.Index
+		if err := json.Unmarshal(content, &index); err != nil {
+			return nil, err
+		}
+		manifests = append(manifests, index.Manifests...)
+		printUpdateStatus(status.IndexPromptMerged, indexRef, string(desc.Digest), opts.Printer)
+	}
+	return manifests, nil
+}
+
+func removeManifests(ctx context.Context, manifests []ocispec.Descriptor, target oras.ReadOnlyTarget, opts updateOptions) ([]ocispec.Descriptor, error) {
+	set := make(map[digest.Digest]int)
+	for _, manifest := range opts.removeArguments {
+		printUpdateStatus(status.IndexPromptResolving, manifest, "", opts.Printer)
+		desc, err := oras.Resolve(ctx, target, manifest, oras.DefaultResolveOptions)
+		if err != nil {
+			return nil, fmt.Errorf("could not resolve the manifest %s: %w", manifest, err)
+		}
+		if !descriptor.IsManifest(desc) {
+			return nil, fmt.Errorf("%s is not a manifest", manifest)
+		}
+		printUpdateStatus(status.IndexPromptResolved, manifest, string(desc.Digest), opts.Printer)
+		set[desc.Digest] = set[desc.Digest] + 1
+	}
+	pointer := len(manifests) - 1
+	for i := len(manifests) - 1; i >= 0; i-- {
+		if _, exists := set[manifests[i].Digest]; exists {
+			val := manifests[i]
+			// move the item to the end of the slice
+			for j := i; j < pointer; j++ {
+				manifests[j] = manifests[j+1]
+			}
+			manifests[pointer] = val
+			pointer = pointer - 1
+			printUpdateStatus(status.IndexPromptRemoved, string(val.Digest), "", opts.Printer)
+			set[val.Digest] = set[val.Digest] - 1
+			if set[val.Digest] == 0 {
+				delete(set, val.Digest)
+			}
+		}
+	}
+	// shrink the slice to remove the manifests
+	manifests = manifests[:pointer+1]
+	for key := range set {
+		return nil, fmt.Errorf("%s does not exist in the index %s", key, opts.Reference)
```

Comment:
This part is a bit hard to read. Can we refactor it into a function and have unit tests for it?